NDTensors/src/UnallocatedArrays/src/unallocatedfill.jl

```diff
@@ -0,0 +1,34 @@
+## TODO All constructors not fully implemented but so far it matches the 
+## constructors found in `FillArrays`. Need to fix io
+struct UnallocatedFill{ElT,N,Axes,Alloc<:AbstractArray} <:
+       FillArrays.AbstractFill{ElT,N,Axes}
+  f::FillArrays.Fill{ElT,N,Axes}
+
+  function UnallocatedFill{ElT,N,Axes,Alloc}(x::ElT, inds::Tuple) where {ElT,N,Axes,Alloc}
+    f = FillArrays.Fill(x, inds)
+    ax = typeof(FillArrays.axes(f))
+    return new{ElT,N,ax,Alloc}(f)
+  end
+
+  function UnallocatedFill{ElT,0,Tuple{},Alloc}(x::ElT, inds::Tuple{}) where {ElT,Alloc}
+    f = FillArrays.Fill{ElT,0,Tuple{}}(x, inds)
+    return new{ElT,0,Tuple{},Alloc}(f)
+  end
+end
+
+function UnallocatedFill{ElT,Alloc}(
+  x::ElT, inds::Tuple
+) where {ElT<:Number,Alloc<:AbstractArray}
+  N = length(inds)
+  Ax = Base.axes(inds)
+  return UnallocatedFill{ElT,N,Ax,Alloc}(x, inds)
+end
```

Comment:
We shouldn't be inconsistent about what parameters mean across constructors, i.e. the second parameter should always be the number of dimensions.

Instead of using a constructor to specify the allocation type, we could use a function `set_alloctype` which makes use of `set_parameters`. It could also convert types like `Zeros` to `UnallocatedZeros`.

Reply:
So then the way to construct `UnallocatedZeros` would be:
```julia
set_alloctype(Zeros(2, 2), Array)
```
The constructor could also specify any unspecified parameters of the allocation type. Internally it could be defined as:
```julia
function set_alloctype(fillarray::AbstractFill, alloctype::Type{<:AbstractArray})
  unallocated_fillarray = unallocated(fillarray)
  return set_alloctype(unallocated_fillarray, alloctype)
end

function set_alloctype(fillarray::AbstractUnallocatedFill, alloctype::Type{<:AbstractArray})
  # Use `set_parameters` to modify the type of `fillarray` based on `alloctype`
end

# Generic conversion from simple FillArray type to an UnallocatedFillArray type
unallocated(fillarray::Zeros) = UnallocatedZeros(fillarray)
```